net/rclone/Makefile

```diff
@@ -0,0 +1,82 @@
+# SPDX-Identifier-License: GPL-3.0-or-later
+#
+# Copyright (C) 2019 Elon Huang <elonhhuang@gmail.com>
+# Copyright (C) 2021 ImmortalWrt.org
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=rclone
+PKG_VERSION:=1.56.2
+PKG_RELEASE:=$(AUTORELEASE)
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/rclone/rclone/tar.gz/v$(PKG_VERSION)?
+PKG_HASH:=a5b0b7dfe17d9ec74e3a33415eec4331c61d800d8823621e61c6164e8f88c567
+
+PKG_LICENSE:=MIT
+PKG_LICENSE_FILE:=LICENSE
+PKG_MAINTAINER:=Elon Huang <elonhhuang@gmail.com> \
+		Tianling Shen <cnsztl@immortalwrt.org>
+
+PKG_BUILD_DEPENDS:=golang/host
+PKG_BUILD_PARALLEL:=1
+PKG_USE_MIPS16:=0
+
+GO_PKG:=github.com/rclone/rclone
+GO_PKG_EXCLUDES:=test
+GO_PKG_LDFLAGS_X:= \
+	github.com/rclone/rclone/fs.Version=v$(PKG_VERSION) \
+	main.Version=v$(PKG_VERSION) \
+	main.BuildUser=openwrt \
+	main.BuildHost=openwrt
+
+include $(INCLUDE_DIR)/package.mk
+include ../../lang/golang/golang-package.mk
+
+define Package/rclone/Default
+  SECTION:=net
+  CATEGORY:=Network
+  SUBMENU:=File Transfer
+  TITLE:=rsync for cloud storage
+  URL:=https://rclone.org
+endef
+
+define Package/rclone
+  $(call Package/rclone/Default)
+  DEPENDS:=$(GO_ARCH_DEPENDS) +ca-bundle +fuse-utils
+  USERID:=rclone:rclone
+endef
+
+define Package/rclone-config
+  TITLE+= (Config Scripts)
+  DEPENDS:=+rclone
+endef
+
+define Package/rclone/description
+  Rclone ("rsync for cloud storage") is a command line program to sync
+  files and directories to and from different cloud storage providers.
+endef
+
+define Package/rclone-config/conffiles
+/etc/config/rclone
+endef
+
+# Disable CGO due to ld linker issue
+GO_PKG_TARGET_VARS:=$(filter-out CGO_ENABLED=%,$(GO_PKG_TARGET_VARS)) CGO_ENABLED=0
```

Comment:
> This error can be avoided by disabling/patching out Go plugins.

Hmm interesting. I didn't really understand why that fixes it - is it a Go compiler toolchain bug?

We could do a build tag for disabling plugins if that would help.

We usually build without CGO on non Windows/macOS platforms so that is an acceptable solution too.

(Rclone maintainer)

Reply:
> Hmm interesting. I didn't really understand why that fixes it - is it a Go compiler toolchain bug?

I don’t think from Go’s perspective this is necessarily their issue, since they assume/require the gold linker is available.

The gold linker isn’t used for anything else in OpenWrt so not having it is not really a bug either.

The Go compiler will [require the gold linker](https://github.com/golang/go/blob/go1.17.3/src/cmd/link/internal/ld/lib.go#L1399) for arm/arm64 when dynamic linking is requested. One of the things that enables dynamic linking is use of Go plugins.

Disabling CGO also bypasses this issue by turning off dynamic linking entirely, which also disables Go plugins.

Hopefully https://github.com/golang/go/issues/22040 can be resolved soon, this should alleviate any need to disable CGO or Go plugins in Go programs.

Reply:
@1715173329 I have added a patch for the Go compiler in #17446 to remove the gold linker requirement. rclone should compile correctly now without having to disable Go plugins.

Reply:
@jefferyto

https://github.com/immortalwrt/packages/runs/4666298378?check_suite_focus=true
https://github.com/immortalwrt/packages/runs/4666298407?check_suite_focus=true

Looks like it is still broken on mips(el).

Reply:
My apologies, I guess I spoke too soon. (Found someone else who ran into this error at https://forum.openwrt.org/t/rclone-on-openwrt/10671/4.) I'll try to find some time to investigate this error soon.